File: .codex/hooks/go_guidelines.py
#!/usr/bin/env python3
"""Resolve module-scoped Go guidance without installing tools during a hook."""

from __future__ import annotations

import argparse
import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
import time
from pathlib import Path

from generated_code_guard import target_paths
from hook_response import additional_context
from locked_state import ensure_private_dir
from session_scope import agent_digest, digest, session_state_dir
from shell_command import shell_segments, strip_leading_shell_prefix


PROVIDER = "use-modern-go/scripts/"
# Leave room for check_project's ordinary detection inside its 10-second hook.
GUIDANCE_TIMEOUT = 6
GO_DIRECTIVE = re.compile(r"^[ \t]*go[ \t]+(\d+\.\d+(?:\.\d+)?)[ \t]*(?://[^\n]*)?$", re.M)
CLI_VERSION = re.compile(r"v\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?")


def module_version(directory: Path) -> tuple[Path | None, str | None]:
    """A workspace/toolchain version is not a member module's language version."""
    for parent in (directory, *directory.parents):
        manifest = parent / "go.mod"
        if manifest.is_file():
            match = GO_DIRECTIVE.search(manifest.read_text(encoding="utf-8"))
            return manifest, match[1] if match else None
    return None, None


def cache_root() -> Path:
    return Path(os.environ.get("XDG_CACHE_HOME") or Path.home() / ".cache")


def atomic_write(path: Path, body: str) -> None:
    ensure_private_dir(path.parent)
    temporary = None
    try:
        with tempfile.NamedTemporaryFile(mode="w", encoding="utf-8", dir=path.parent,
                                         prefix=".tmp-", delete=False) as stream:
            temporary = Path(stream.name)
            stream.write(body)
        temporary.replace(path)
    finally:
        if temporary is not None:
            temporary.unlink(missing_ok=True)


def guideline_list(manifest: Path, version: str, deadline: float) -> tuple[str, str]:
    def run(command: list[str]) -> str:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise subprocess.TimeoutExpired(command, GUIDANCE_TIMEOUT)
        result = subprocess.run(command, cwd=manifest.parent, capture_output=True,
                                text=True, timeout=remaining,
                                check=False)
        if result.returncode or not result.stdout.strip():
            raise RuntimeError("the guidelines provider returned no usable result")
        return result.stdout.strip()

    manager = shutil.which("skills-mgr")
    if manager is None:
        raise RuntimeError("skills-mgr is unavailable")
    cli_version = run([manager, "get", PROVIDER + "VERSION"])
    if not CLI_VERSION.fullmatch(cli_version):
        raise RuntimeError("the guidelines provider returned an invalid CLI version")

    # run-tool.sh owns this versioned cache layout and installs on a cache miss.
    # Run its already-installed binary directly so a startup hook cannot install.
    binary = cache_root() / "go-modern-guidelines" / cli_version / "go-modern-guidelines"
    if not binary.is_file() or not os.access(binary, os.X_OK):
        raise RuntimeError(f"Modern Go Guidelines {cli_version} is not installed; ask before installing it")

    cached = cache_root() / "codex" / "go-guidelines" / "lists" / (digest(cli_version, version) + ".txt")
    try:
        body = cached.read_text(encoding="utf-8")
        if body.strip():
            return cli_version, body
    except OSError:
        pass
    if run([str(binary), "--version"]) != cli_version:
        raise RuntimeError("the installed guidelines CLI does not match the provider version")
    body = run([str(binary), "list", "--go-version", version])
    try:
        atomic_write(cached, body)
    except OSError:
        # A read-only cache must not prevent delivering a successfully generated list.
        pass
    return cli_version, body


def report(directory: Path, languages: str, deadline: float) -> tuple[str, bool]:
    try:
        manifest, version = module_version(directory)
    except (OSError, UnicodeError):
        manifest, version = None, None
        unreadable = True
        error = "the owning go.mod could not be read"
    else:
        unreadable = False
        error = "select the target module and resolve its go directive; the installed toolchain is not a substitute"

    fields = [f"go_version: {version or 'unknown'}",
              "go_module: " + json.dumps(str(manifest) if manifest else None)]
    relevant = unreadable or manifest is not None or "go" in languages.split(",") or (directory / "go.work").is_file()
    status = "not-applicable"
    body = ""
    if version is None:
        if relevant:
            status = "unavailable" if unreadable else "unresolved"
            body = f"Go guidance is not loaded: {error}. Hooks update guidance automatically as the target becomes known."
    else:
        try:
            cli_version, guidelines = guideline_list(manifest, version, deadline)
        except subprocess.TimeoutExpired:
            status = "unavailable"
            body = "Go guidance is not loaded: the provider timed out. Hooks retry automatically; report a persistent blocker before Go work."
        except (OSError, UnicodeError, RuntimeError) as exc:
            status = "unavailable"
            body = f"Go guidance is not loaded: {exc}. Hooks retry automatically; report a persistent blocker before Go work."
        else:
            status = "ready"
            body = (
                f"Modern Go Guidelines {cli_version}, scoped to {manifest} (Go {version}).\n"
                "For Go writing, refactoring, review, or testing, the only initial skill read is "
                "`skills-mgr get golang-best-practices`. use-modern-go is the CLI backend; its list is already supplied here.\n"
                "Read the complete list below, newest first. Apply relevant guidelines even when nearby code uses older idioms. "
                "Skip only when a rule would not compile, would change behavior, or does not fit the edited code.\n"
                "Before skipping a seemingly relevant rule, or when examples are needed, request only its returned IDs with "
                "`skills-mgr run use-modern-go/scripts/run-tool.sh explain <ID> [<ID> ...]`.\n\n"
                + guidelines
            )
    fields.append("go_guidelines_status: " + status)
    if body:
        fields.append("go_guidelines: |\n" + "\n".join("  " + line for line in body.splitlines()))
    return "\n".join(fields) + "\n", status != "unavailable"


def target_directories(event: dict, cwd: Path) -> tuple[Path, ...]:
    """Use paths already carried by normal tool calls, not an agent refresh call."""
    tool_input = event.get("tool_input")
    if isinstance(tool_input, dict):
        for field in ("workdir", "cwd"):
            value = tool_input.get(field)
            if isinstance(value, str) and value:
                cwd = (cwd / Path(value).expanduser()).resolve()
                break
    paths = target_paths(dict(event, cwd=str(cwd)))
    directories = [path.parent for path in paths
                   if path.suffix == ".go" or path.name in {"go.mod", "go.work"}]
    if isinstance(tool_input, dict):
        for field in ("command", "cmd"):
            command = tool_input.get(field)
            if not isinstance(command, str):
                continue
            shell_cwd = cwd
            for segment in shell_segments(command):
                tokens = strip_leading_shell_prefix(segment)
                if len(tokens) >= 2 and tokens[0] == "cd":
                    destination = tokens[-1]
                    if not destination.startswith("-"):
                        shell_cwd = (shell_cwd / Path(destination).expanduser()).resolve()
                        directories.append(shell_cwd)
                    continue
                for token in tokens:
                    try:
                        path = shell_cwd / Path(token).expanduser()
                        if path.suffix == ".go" or path.name in {"go.mod", "go.work"}:
                            directories.append(path.parent)
                        elif path.is_dir():
                            directories.append(path)
                    except (OSError, RuntimeError, ValueError):
                        # Shell arguments can be code or expressions, not paths.
                        continue
    # Many inspected files or patch targets can belong to one module. Resolve
    # those before invoking the provider so they share one request and budget.
    modules: dict[Path, None] = {}
    for directory in directories or [cwd]:
        directory = directory.resolve()
        try:
            manifest, _ = module_version(directory)
        except (OSError, UnicodeError):
            manifest = None
        modules[manifest.parent if manifest else directory] = None
    return tuple(modules)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("directory", nargs="?")
    parser.add_argument("--languages", default="none")
    parser.add_argument("--refresh", action="store_true", help="inject only when the module guidance changes")
    args = parser.parse_args()
    event = {}
    if not sys.stdin.isatty():
        try:
            value = json.load(sys.stdin)
            if isinstance(value, dict):
                event = value
        except (ValueError, OSError):
            pass
    directory = Path(args.directory or event.get("cwd") or os.getcwd()).resolve()
    directories = target_directories(event, directory) if args.refresh else (directory,)
    deadline = time.monotonic() + GUIDANCE_TIMEOUT
    reports = [report(target, args.languages, deadline) for target in directories]
    body = "\n".join(dict.fromkeys(body for body, _ in reports))
    reusable = all(reusable for _, reusable in reports)
    state_dir = session_state_dir(event, cache_root() / "codex" / "go-guidelines" / "contexts")
    state = state_dir / (agent_digest(event) or "root") if state_dir else None
    fingerprint = digest(body)
    if args.refresh and state is not None:
        try:
            if reusable and state.read_text(encoding="utf-8") == fingerprint:
                return 0
        except OSError:
            pass
    if args.refresh:
        event_name = event.get("hook_event_name", "UserPromptSubmit")
        print(json.dumps(additional_context(body, event_name=event_name)))
    else:
        print(body, end="")
    # Delivery state describes the latest context, including a failure. Otherwise
    # recovery to an earlier ready report would be mistaken for an unchanged one.
    if state is not None:
        try:
            atomic_write(state, fingerprint)
        except OSError:
            pass
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .codex/hooks/go_guidelines.ts
import { createHash } from "crypto";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { additionalContext } from "./lib/hook_response.ts";
import { asString, handleVersion, isRecord, readEvent, runCommand, runMain, writeJson } from "./lib/hook_runtime.ts";
import { shellTokens } from "./lib/shell_command.ts";

export const VERSION = "1.1.0";

// Preserve quoting until after splitting: a quoted semicolon/newline is an
// argument, not evidence that a separate skill command ran.
function batchCommands(command: string): { text: string; separator: string }[] | undefined {
  const commands: { text: string; separator: string }[] = [];
  let text = "";
  let quote = "";
  let boundary = true;
  for (let index = 0; index < command.length; index += 1) {
    const character = command[index];
    if (character === "\\" && quote !== "'") {
      if (index + 1 >= command.length) return undefined;
      if (command[index + 1] !== "\n") {
        text += character + command[index + 1];
        boundary = false;
      }
      index += 1;
      continue;
    }
    if (quote) {
      if (character === quote) quote = "";
      // Expansions can execute commands and change the meaning of arguments.
      else if (quote === '"' && (character === "$" || character === "`")) return undefined;
      text += character;
      continue;
    }
    if (character === "'" || character === '"') {
      quote = character;
      text += character;
      boundary = false;
      continue;
    }
    if (character === "#" && boundary) {
      while (index < command.length && command[index] !== "\n") index += 1;
      if (index === command.length) break;
    } else if ("|()<>{}$`".includes(character)) {
      return undefined;
    }
    let separator = command[index];
    if (separator === "&") {
      if (index + 1 >= command.length || command[index + 1] !== "&") return undefined;
      separator = "&&";
    }
    if (separator === "\n" || separator === ";" || separator === "&&") {
      commands.push({ text, separator });
      text = "";
      boundary = true;
      if (separator === "&&") index += 1;
    } else {
      text += character;
      boundary = /[ \t\r]/.test(character);
    }
  }
  if (quote) return undefined;
  commands.push({ text, separator: "" });
  return commands;
}

function isRecordOrEmpty(value: unknown): value is Record<string, unknown> {
  return isRecord(value);
}

export function skillDirectory(event: Record<string, unknown>): string | undefined {
  if (event.hook_event_name !== "PostToolUse") {
    return undefined;
  }
  const result = event.tool_response;
  if (isRecordOrEmpty(result)) {
    const details = isRecord(result.details) ? result.details : {};
    const exitCode = result.exit_code;
    const asyncState = isRecord(details.async) ? details.async.state : undefined;
    if (
      result.is_error ||
      (exitCode !== undefined && exitCode !== null && exitCode !== 0) ||
      details.timedOut ||
      asyncState === "running" ||
      asyncState === "failed"
    ) {
      return undefined;
    }
  }
  const tool = isRecord(event.tool_input) ? event.tool_input : {};
  const command = tool.command ?? tool.cmd;
  if (typeof command !== "string") {
    return undefined;
  }
  let cwd = asString(event.cwd) || process.cwd();
  const workdir = asString(tool.workdir) || asString(tool.cwd) || ".";
  cwd = path.resolve(cwd, workdir.startsWith("~/") ? path.join(os.homedir(), workdir.slice(2)) : workdir);
  const parsed = batchCommands(command);
  if (parsed === undefined) return undefined;
  const commands = parsed.filter((part) => shellTokens(part.text).length > 0);
  // A later unconditional command can hide a failed/skipped && chain.
  // Accept a complete success chain or unconditional reads, not a mix.
  if (commands.some((part) => part.separator === "&&") &&
      commands.slice(0, -1).some((part) => part.separator !== "&&")) return undefined;
  let directory: string | undefined;
  for (const part of commands) {
    let tokens = shellTokens(part.text);
    if (tokens.length === 0) continue;
    const first = tokens[0];
    // Only straight-line batches: overall success cannot prove execution
    // through branches, loops, shell replacement, or directory-stack changes.
    if (/^[A-Za-z_][A-Za-z0-9_]*=/.test(first) || [
      "if", "then", "else", "elif", "fi", "for", "while", "until", "do", "done",
      "case", "esac", "function", "select", "repeat", "coproc", "!", "exit", "return",
      "exec", "eval", "source", ".", "pushd", "popd", "set", "trap", "command", "builtin",
    ].includes(first)) return undefined;
    if (first === "cd") {
      if (tokens.length !== 2 || part.separator !== "&&") return undefined;
      const target = tokens[1];
      if (!target || target.startsWith("-") || /[$`*?\[\]]/.test(target)) return undefined;
      cwd = path.resolve(cwd, target.startsWith("~/") ? path.join(os.homedir(), target.slice(2)) : target);
      continue;
    }
    if (path.basename(first) === "rtk") tokens = tokens.slice(1);
    if (tokens.length < 3 || path.basename(tokens[0]) !== "skills-mgr" || tokens[1] !== "get") continue;
    const args = tokens.slice(2).filter((token) => !["--codex", "--claude", "--grok"].includes(token));
    if (args.length !== 1 || !["golang-best-practices", "golang-best-practices/SKILL.md"].includes(args[0])) continue;
    if (directory !== undefined && directory !== cwd) return undefined;
    directory = cwd;
  }
  return directory;
}

function findGoMod(directory: string): string | undefined {
  let current = directory;
  while (true) {
    const candidate = path.join(current, "go.mod");
    if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
      return candidate;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      return undefined;
    }
    current = parent;
  }
}

export function guidance(directory: string): string {
  const manifest = findGoMod(directory);
  if (manifest === undefined) {
    throw new Error("no owning go.mod; invoke the skill from the target module");
  }
  const text = fs.readFileSync(manifest, "utf8");
  const match = /^\s*go[ \t]+(\d+\.\d+(?:\.\d+)?)[ \t]*(?:\/\/[^\n]*)?$/m.exec(text);
  if (match === null) {
    throw new Error("the owning go.mod has no valid go directive");
  }
  const version = match[1] ?? "";
  const providerResult = runCommand(["skills-mgr", "get", "use-modern-go/scripts/VERSION"], {
    cwd: directory,
  });
  if (providerResult.status !== 0) {
    throw new Error(providerResult.stderr.trim() || "skills-mgr VERSION lookup failed");
  }
  const provider = providerResult.stdout.trim();
  if (!/^v\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?$/.test(provider)) {
    throw new Error("invalid guidelines provider version");
  }
  const cache = process.env.XDG_CACHE_HOME || path.join(os.homedir(), ".cache");
  const binary = path.join(cache, "go-modern-guidelines", provider, "go-modern-guidelines");
  try {
    fs.accessSync(binary, fs.constants.X_OK);
  } catch {
    throw new Error(`Modern Go Guidelines ${provider} is missing; ask before installing`);
  }
  const listed = runCommand([binary, "list", "--go-version", version], {
    cwd: directory,
  });
  if (listed.status !== 0) {
    throw new Error(listed.stderr.trim() || "guidelines provider failed");
  }
  const body = listed.stdout.trim();
  if (body.length === 0) {
    throw new Error("the guidelines provider returned an empty list");
  }
  const digest = createHash("sha256").update(body).digest("hex");
  return `Modern Go Guidelines ${provider}: ${manifest} (Go ${version})\n${body}\nEND_GO_GUIDELINES sha256=${digest}`;
}

export function responseFor(event: unknown): Record<string, unknown> | undefined {
  if (!isRecord(event)) {
    return undefined;
  }
  const directory = skillDirectory(event);
  if (directory === undefined) {
    return undefined;
  }
  let body: string;
  try {
    body = guidance(directory);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    body = `Go guidelines unavailable: ${message}. Report this blocker before Go work.`;
  }
  return additionalContext(body, "PostToolUse");
}

function main(): number {
  if (handleVersion(VERSION)) {
    return 0;
  }
  const response = responseFor(readEvent());
  if (response !== undefined) {
    writeJson(response);
  }
  return 0;
}

runMain("go_guidelines", main);
